File: laravel-backend/app/Utils/RateLimiter.php
<?php

namespace App\Utils;

class RateLimiter
{
    private int $maxRequests;
    private int $timeWindow;
    private array $requests = [];

    public function __construct(int $maxRequests, int $timeWindowSeconds)
    {
        $this->maxRequests = $maxRequests;
        $this->timeWindow = $timeWindowSeconds;
    }

    public function acquire(): void
    {
        $now = microtime(true);
        $this->requests = array_filter($this->requests, fn($t) => $now - $t < $this->timeWindow);

        if (count($this->requests) >= $this->maxRequests) {
            $sleepTime = $this->timeWindow - ($now - $this->requests[0]);
            if ($sleepTime > 0) {
                usleep((int) ($sleepTime * 1_000_000));
            }
        }

        $this->requests[] = microtime(true);
    }
}

File: backend/app/utils/rate_limiter.py
import asyncio
import time


class RateLimiter:
    """Token bucket rate limiter."""

    def __init__(self, max_requests: int, time_window_seconds: int):
        self.max_requests = max_requests
        self.time_window = time_window_seconds
        self.requests: list[float] = []

    async def acquire(self):
        """Wait until a request slot is available."""
        now = time.time()
        self.requests = [t for t in self.requests if now - t < self.time_window]
        if len(self.requests) >= self.max_requests:
            sleep_time = self.time_window - (now - self.requests[0])
            if sleep_time > 0:
                await asyncio.sleep(sleep_time)
        self.requests.append(time.time())

    def acquire_sync(self):
        """Synchronous version for sync adapters."""
        now = time.time()
        self.requests = [t for t in self.requests if now - t < self.time_window]
        if len(self.requests) >= self.max_requests:
            sleep_time = self.time_window - (now - self.requests[0])
            if sleep_time > 0:
                time.sleep(sleep_time)
        self.requests.append(time.time())
